crates/rust-analyzer/src/lsp_ext.rs

```diff
@@ -97,6 +97,23 @@ pub struct JoinLinesParams {
     pub ranges: Vec<Range>,
 }
 
+pub enum ResolveCodeActionRequest {}
+
+impl Request for ResolveCodeActionRequest {
+    type Params = ResolveCodeActionParams;
+    type Result = Option<SnippetWorkspaceEdit>;
+    const METHOD: &'static str = "experimental/resolveCodeAction";
+}
+
+/// Params for the ResolveCodeActionRequest
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ResolveCodeActionParams {
+    pub code_action_params: lsp_types::CodeActionParams,
+    pub id: String,
+    pub label: String,
```

Comment:
Can we remove this?

Reply:
I can, but I just want to determine the right strategy. Can you please look at my last comment here: https://github.com/rust-analyzer/rust-analyzer/pull/4576#issuecomment-637763634 and maybe reply to it.

Reply:
Right, I think I was looking for that comment in this PR ;) 



> What shall we do with diagnostics? 

Just compute edits eagarly is probably the right call for now. 

> Shall we in resolveCodeAction call all the assists or only the specific one? 

I think `fn resolve_assist(&self, range: FileRange, id: AsssitId, disambiguator: usize) -> Option<ResolvedAssist>` would be slightly better, eventually, but at the moment just computing them all matches the current implementaiton better. Long-term, I think we should change something around `AssistCtx` to avoid even computing the assistrs with wrong IDs, but that's not too important I would think. So I'd keep it simple and just use the current API

> You are right that for groups we cannot just use the AssistId, we also need to add the indicator which will indicate the option in the group for example.

I think the simplest and most reliable solution is to just guarantee that assists are produced in the same order. Actually, for the current impelentation we can even just remove `AssistId` completely and just use index to directly fetch the assist. Let's maybe do this? Just always encode id as `{assist_id}:{index}`, and then, at `resolve_assist`, `assert!` that IDs match. We *should* implement smarter id filtering at some point, but it's usually better to postpone everything you can postepone :)